core/src/main/scala/kafka/server/DelayedOperation.scala

```diff
@@ -168,11 +170,25 @@ final class DelayedOperationPurgatory[T <: DelayedOperation](purgatoryName: Stri
                                                              reaperEnabled: Boolean = true,
                                                              timerEnabled: Boolean = true)
         extends Logging with KafkaMetricsGroup {
-
   /* a list of operation watching keys */
-  private val watchersForKey = new Pool[Any, Watchers](Some((key: Any) => new Watchers(key)))
+  private class WatcherList {
+    val watchersByKey = new Pool[Any, Watchers](Some((key: Any) => new Watchers(key)))
+
+    val watchersLock = new ReentrantLock()
+
+    /*
+     * Return all the current watcher lists,
+     * note that the returned watchers may be removed from the list by other threads
+     */
+    def allWatchers = {
+      inLock(watchersLock) { watchersByKey.values }
```

Comment:
This is an existing issue. But I am wondering if we really need the lock here. What we return from the lock is a view of the backing map, which can change after the lock is released. Since ConcurrentHashMap supports weakly consistent iterator already, it seems that we don't need the lock.